---
title: "Import and representation of Mouse Aging Data"
author: 
    - name: Tram Nguyen
      affiliation: Center for Computational Biomedicine, Harvard Medical School
      email: Tram_Nguyen@hms.harvard.edu
    - name: Kris Holton
      affiliation: Harvard Stem Cell Institute, Harvard Medical School
    - name: Tyrone Lee
      affiliation: Center for Computational Biomedicine, Harvard Medical School
    - name: Ludwig Geistlinger
      affiliation: Center for Computational Biomedicine, Harvard Medical School
    - name: Robert Gentleman
      affiliation: Center for Computational Biomedicine, Harvard Medical School
package: MouseAgingData
output:
    BiocStyle::html_document:
      self_contained: yes 
      toc: true
      toc_float: true
      toc_depth: 2
      code_folding: show
date: "`r doc_date()`"
vignette: >
    % \VignetteIndexEntry{Mouse ileum}
    %\VignetteEncoding{UTF-8}
    % \VignetteEngine{knitr::rmarkdown}
editor_options: 
    markdown: 
      wrap: 80
---

```{r style, echo = FALSE, results = 'asis', include = FALSE}
  BiocStyle::markdown()
```

# Installation

Install the package using Bioconductor. Start R and enter:

```{r, eval = FALSE}
if(!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("MouseAgingData")
```

# Setup

Now, load the package and dependencies used in the vignette.

```{r, setup, message = FALSE}
library(DropletUtils)
library(edgeR)
library(Seurat)
library(scran)
library(ggplot2)
library(SingleCellExperiment)
library(ExperimentHub)
```

# Data

The `MouseAgingData` package provides analysis-ready data resources of different
aging omics datasets, including the aging mouse brain single cell study
Ximerakis et al., [2019](https://www.nature.com/articles/s41593-019-0491-3),
aging mouse brain parabiosis single cell study Ximerakis et al.,
[2023](https://pubmed.ncbi.nlm.nih.gov/37118429/). In addition, two relevant
transcriptomic studies of rejuvenation are also provided: an aging mouse caloric
restriction single cell study Ma et al.,
[2020](https://pubmed.ncbi.nlm.nih.gov/32109414/), and an aging mouse exercise
single cell study Liu et al., [2022](https://pubmed.ncbi.nlm.nih.gov/37080206/).
The contents of the package can be accessed by querying ExperimentHub with the
package name.

This vignette performs a simple analysis of the parabiosis 10X Genomics
single-cell dataset from Ximerakis et al.
([2023](https://www.nature.com/articles/s43587-023-00373-6)). The analysis
follows the Basics of Single-Cell Analysis in the [OSCA Bioconductor book Quick
Start
Workflow](https://bioconductor.org/books/3.15/OSCA.intro/analysis-overview.html#quick-start-simple).
Steps in this vignette include: quality control, normalization, various forms of
dimensionality reduction, clustering into subpopulations, detection of marker
genes, and annotation of cell types.

```{r}
eh <- ExperimentHub()
query(eh, c("MouseAgingData", "parabiosis10X"))
```

The individual data sets can be accessed using either their ExperimentHub
accession number, or the provided functions in this package. For example, to
load the parabiosis 10X Genomics droplet data from Ximerakis et al. 2023, do the
following:

```{r}
droplet <- eh[["EH1617"]]
droplet
droplet <- parabiosis10X()
droplet
```

To view a simple workflow to analyze and visualize the downloaded data following
the [OSCA Single-Cell
Workbook](https://bioconductor.org/books/3.18/OSCA.basic/), do the following:

```{r}
browseVignettes("MouseAgingData")
```

# Session info

```{r}
sessionInfo()
```
