---
title: "Import and representation of parabiosis droplet data"
author: 
    - name: Tram Nguyen
      affiliation: Center for Computational Biomedicine, Harvard Medical School
      email: Tram_Nguyen@hms.harvard.edu
    - name: Kris Holton
      affiliation: Harvard Stem Cell Institute, Harvard Medical School
    - name: Tyrone Lee
      affiliation: Center for Computational Biomedicine, Harvard Medical School
    - name: Ludwig Geistlinger
      affiliation: Center for Computational Biomedicine, Harvard Medical School
    - name: Robert Gentleman
      affiliation: Center for Computational Biomedicine, Harvard Medical School
package: MouseAgingData
output:
    BiocStyle::html_document:
      self_contained: yes 
      toc: true
      toc_float: true
      toc_depth: 2
      code_folding: show
date: "`r doc_date()`"
vignette: >
    % \VignetteIndexEntry{parabiosis droplet}
    %\VignetteEncoding{UTF-8}
    % \VignetteEngine{knitr::rmarkdown}
editor_options: 
    markdown: 
      wrap: 80
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>",
    crop = NULL
)
```

# Installation

Install the package using Bioconductor. Start R and enter:

```{r, eval = FALSE}
# if(!requireNamespace("BiocManager", quietly = TRUE))
    # install.packages("BiocManager")
# BiocManager::install("MouseAgingData")
```

# Setup

Now, load the package and dependencies used in the vignette.

```{r, message = FALSE}
library(scran)
library(ggplot2)
library(SingleCellExperiment)
library(ExperimentHub)
```

# Introduction

Single-cell sequencing technology can reveal intricate details about individual
cells, allowing researchers to interrogate the genetic make up of cells within a
heterogeneous sample. Single-cell sequencing can provide insights into various
aspects of cellular biology, such as characterization of cell populations,
identification of rare cell types, and quantification of expression levels in
cell types across experimental treatments. Given the wide utility, single-cell
sequencing has expanded scientific knowledge various fields, including cancer
research, immunology, developmental biology, neurobiology, and microbiology.

There are several methods for generating single-cell sequencing data which can
extract either the DNA (scDNA-seq) or RNA (scRNA-seq) of a cell. These include,
but are not limited to:

1. Droplet-based platforms such as 10x Genomics Chromium system, inDrop,
Drop-seq, and Seq-Well, which use microfluidic devices to isolate individual
cells into tiny droplets along with unique barcoded beads.

2. Plate or microwell-based methods such as the Smart-seq2 protocol or the C1
system by Fluidigm, respectively. These platforms employ microfluidic chips or
multi-well arrays to capture and process individual cells. Unlike droplet-based
platforms, these cells are manually or automatically sorted into individual
wells of the plate.

The `MouseAgingData` package provides analysis-ready data from an aging mouse
brain parabiosis single cell study by Ximerakis & Holton et al.,
[2023](https://pubmed.ncbi.nlm.nih.gov/37118429/). The contents of the package
can be accessed by querying ExperimentHub with the package name.


# Data

This vignette performs a simple analysis of the parabiosis 10X Genomics
single-cell data set from Ximerakis & Holton et al.
([2023](https://www.nature.com/articles/s43587-023-00373-6)). The analysis
follows the Basics of Single-Cell Analysis in the [OSCA Bioconductor book Quick
Start
Workflow](https://bioconductor.org/books/3.15/OSCA.intro/analysis-overview.html#quick-start-simple).
In this study, the authors investigated how heterochronic parabiosis affects the
mouse brain in terms of aging and rejuvenation. They identified gene signatures
attributed to aging in specific cell-types. They focus especially on brain
endothelial cells, which showed dynamic transcriptional changes that affect
vascular structure and function.

This vignette walks through the process of quality control, normalization,
various forms of dimensionality reduction, clustering, detection of marker
genes, and annotation of cell types. PCA, UMAP, and tSNE coordinates used in the
study were provided b the authors for visualization.

The parabiosis single cell RNA-seq (Ximerakis, Holton et al Nature Aging 2023)
includes 105,329 cells, 31 cell types across 8 OX, 8 YX, 7 YY, 9 YO, 7 OO, 11 OY
animals, and 20905 features


# Load the data set

```{r, eval = FALSE}
eh <- ExperimentHub()
query(eh, c("MouseAgingData", "parabiosis10X"))

# Eventually we can load in data through our package, for now:
sce <- readRDS("~/parabiosis_sce.rds")
sce
```

Do some quick QC to make sure the data loaded correctly and is what we expected.

```{r, Data check}

sce <- readRDS("~/parabiosis_sce.rds")
sce

# Check sce dimensions
dim(sce)

# Sample metadata
head(colData(sce)) 

# Includes cell colors from the original paper
metadata(sce)

# The count matrix. Transcript for each gene per cell (column)
head(counts(sce)) 
```

# Quality control

In this step, we can explore and visualize mitochondrial content and read count.
However, the authors have already removed low-quality cells and animals so we
will skip this section in this vignette. For more details on their workflow, one
can refer to the original article Ximerakis & Holton et al.
([2023](https://www.nature.com/articles/s43587-023-00373-6)). The [OSCA
Bioconductor
book](https://bioconductor.org/books/3.15/OSCA.basic/quality-control.html) also
provides several examples of quality control steps as well.

# Normalization

```{r, normalize}
library(scran)
set.seed(101000110)
clusters <- quickCluster(sce)
sce <- computeSumFactors(sce, clusters=clusters)
sce <- logNormCounts(sce)


head(logcounts(sce))
```

# Feature selection

At this point a typical workflow, we could select an appropriate set of highly
variable genes (HVGs), say the top 10% of genes with the highest variability in
expression. Below is an example of how to do this.

```{r, feat select}
library(scran)
dec <- modelGeneVar(sce)
hvg <- getTopHVGs(dec, prop=0.1)
```

A logical index showing the 2000 HVG included in the original study can also be
accessed in the  `SingleCellExperiment` object in the `rowData()` slot.


# PCA

Below is a method for running a Principal Components Analysis using our
previously defined HVGs.

```{r, PCA, warning=FALSE, eval=FALSE}
library(scater)
set.seed(1234)
sce <- runPCA(sce, ncomponents=25, subset_row=hvg) 
Add PCA coords to sce object
```

However, the authors have also provided us with the first 50 PCs used in their
study. For consistent with their study, we will use these.

```{r, viewPCA, warning=FALSE}
head(reducedDim(sce, "PCA"))[1:5, 1:5]
```

# Clustering

```{r, clustering}
library(bluster)
colLabels(sce) <- clusterCells(sce, use.dimred='PCA',
                               BLUSPARAM=NNGraphParam(cluster.fun="louvain"))
```

# Visualization

For this dataset, the authors have already provided us with their exact UMAP and
TSNE coordinates, as well as their color scheme representing the cell types from
their paper. This can be accessed in the metadata slot of the
`SingleCellExperiment` object with the `metadata()` function.

```{r provided umap, fig.height = 4, fig.width = 6.5}
# Generate color map matching cell type to colors in publication
cell.color <- metadata(sce)$cell_color

library(scater)
gg <- plotUMAP(sce, color_by = "cell_type", text_by = "cell_type") 
gg + theme(legend.title=element_blank()) + 
    scale_color_manual(values=c(cell.color))
```

This plot is a recreation of Fig. 2C from Ximerakis & Holton et al. 2023.

We can also plot a TSNE with their provided coordinates.

```{r, plot provided TSNE, fig.height = 4, fig.width = 6.5}
gg <- plotTSNE(sce, color_by = "cell_type", text_by = "cell_type") 
gg + theme(legend.title=element_blank()) + 
    scale_color_manual(values=c(cell.color))
```

<br>

If you would like to create your own UMAP and TSNE plots, please refer back to
the [OSCA Bioconductor book](https://bioconductor.org/books/3.15/OSCA.basic/)
and save our own coordinates to the `SingleCellExperiment` object. Below is some
example code to achieve this.

Start with a UMAP.

```{r, UMAP2, fig.height = 4, fig.width = 6.5}
# Add our UMAP to the reducedDims slot with the name "osca_UMAP" so we do not
# confused it with the author's UMAP coords.
sce <- runUMAP(sce, dimred = 'PCA', name = "osca_UMAP")

# When we want to access a specific item (here, "osca_UMAP") in the reducedDims
# element of the SingleCellExperiment object, we can use the function
# plotReducedDim with the argument dimred.

gg <- plotReducedDim(sce, dimred = "osca_UMAP", color_by = "cell_type", 
                     text_by = "cell_type") 
gg + theme(legend.title=element_blank()) + 
    scale_color_manual(values=c(cell.color))
```

Now with a TSNE.

```{r, TSNE2, fig.height = 4, fig.width = 6.5}
# Run TSNE
set.seed(00101001101)

# runTSNE() stores the t-SNE coordinates in the reducedDims
sce <- runTSNE(sce, dimred = 'PCA', name = "osca_TSNE")

gg <- plotReducedDim(sce, dimred = "osca_TSNE", color_by = "cell_type", 
                     text_by = "cell_type") 
gg + theme(legend.title=element_blank()) + 
    scale_color_manual(values=c(cell.color))
```

# Reference

Ximerakis & Holton et al. (2023) Heterochronic parabiosis reprograms the mouse
brain transcriptome by shifting aging signatures in multiple cell types.
\emph{Nat Aging} 3, 327â€“345. <DOI:https://doi.org/10.1038/s43587-023-00373-6>.

# Session Info

```{r, sesh info}
sessionInfo()
```
