---
title: "Quick OSCA workflow on Parabiosis Mouse Data"
author: "Tram N. Nguyen"
date: "19 December, 2023"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Quick OSCA workflow on Parabiosis Mouse Data}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  tidy.opts = list(width.cutoff = 100), tidy = TRUE
)
```

```{r, setup, message=FALSE}
library(devtools)
library(DropletUtils)
library(edgeR)
library(Seurat)
library(scran)
library(ggplot2)
library(SingleCellExperiment)
library(AnnotationHub)
```

## 1. Introduction

This vignette performs a simple analysis of the Ximerakis et al.
([2023](https://www.nature.com/articles/s43587-023-00373-6)) 10X Genomics
parabiosis dataset. The analysis follows the Basics of Single-Cell Analysis in
the [OSCA Bioconductor book Quick Start
Workflow](https://bioconductor.org/books/3.15/OSCA.intro/analysis-overview.html#quick-start-simple).
Steps in this vignette include: quality control, normalization, various forms of
dimensionality reduction, clustering into subpopulations, detection of marker
genes, and annotation of cell types.


## 2. Data Loading

### About the dataset

Parabiosis single cell RNA-seq (Ximerakis, Holton et al Nature Aging 2023)
105,329 cells, 31 cell types
8 OX, 8 YX, 7 YY, 9 YO, 7 OO, 11 OY
20905 features

Processed:
(ALL) reduced to: OLG, OPC, GLUT, GABA, EC, ASC
Aging OX v YX; RJV: OY v OX; AGA YO v YX

<br>

### Make the `SingleCellExperiment` object. 

Load cell barcodes, genes, and matrix files using "DropletUtils".

```{r, Data Loading}
library(MouseAgingData)

# getWebData(format = "tsv", destination = "~/Desktop/", url = "https://ftp.ncbi.nlm.nih.gov/geo/series/GSE222nnn/GSE222510/suppl/GSE222510_processed_barcodes.tsv.gz", header=F)

# Eventually we can load in data through our package, for now:
sce <- readRDS("~/Desktop/parabiosis_sce.rds")
sce
```

Do some quick QC to make sure the data loaded correctly and is what we expected.

```{r, Data check}
# Check sce dimensions
dim(sce)

# Sample metadata
head(colData(sce)) 

# Includes cell colors from the original paper
metadata(sce)

# The count matrix. Transcript for each gene per cell (column)
head(counts(sce)) 
```


## 3. Quality control

In this step, we can explore and visualize mitochondrial content and read count. However, the authors have already removed low-quality cells and animals so we will skip this section in this vignette. For more details on their workflow, one can refer to the original article Ximerakis et al. ([2023](https://www.nature.com/articles/s43587-023-00373-6)). The [OSCA Bioconductor book ](https://bioconductor.org/books/3.15/OSCA.basic/quality-control.html) also provides several examples of quality control steps as well. 



## 4. Normalization

```{r, normalize}
library(scran)
set.seed(101000110)
clusters <- quickCluster(sce)
sce <- computeSumFactors(sce, clusters=clusters)
sce <- logNormCounts(sce)


head(logcounts(sce))
```

## 5. Feature selection

Select an appropriate set of highly variable genes (HVGs).


```{r, feat select}
library(scran)
dec <- modelGeneVar(sce)
hvg <- getTopHVGs(dec, prop=0.1)
```


## 6. PCA

```{r, PCA, warning=F}
library(scater)
set.seed(1234)
sce <- runPCA(sce, ncomponents=25, subset_row=hvg) # Add PCA coords to sce object
```


## 7. Clustering

```{r, clustering}
library(bluster)
colLabels(sce) <- clusterCells(sce, use.dimred='PCA',
                               BLUSPARAM=NNGraphParam(cluster.fun="louvain"))
```


## 8. Visualization

For this dataset, the authors have already provided us with their exact UMAP and TSNE coordinates, as well as their color scheme representing the cell types from their paper. This can be accessed in the metadata slot of the `SingleCellExperiment` object with the `metadata()` function.

```{r provided umap, fig.height = 4, fig.width = 6.5}
# Generate color map matching cell type to colors in publication
cell.color <- metadata(sce)$cell_color

library(scater)
gg <- plotUMAP(sce, color_by = "cell_type", text_by = "cell_type") 
gg + theme(legend.title=element_blank()) + scale_color_manual(values=c(cell.color))
```

This plot is a recreation of Fig. 2C from Ximerakis et al. 2023.


We can also plot a TSNE with their provided coordinates.

```{r, plot provided TSNE, fig.height = 4, fig.width = 6.5}
gg <- plotTSNE(sce, color_by = "cell_type", text_by = "cell_type") 
gg + theme(legend.title=element_blank()) + scale_color_manual(values=c(cell.color))
```

<br>

We can also create our own UMAP and TSNE plots following the OSCA workflow and save our own coordinates to the `SingleCellExperiment` object. 


Start with a UMAP.
```{r, UMAP2, fig.height = 4, fig.width = 6.5}
# Add our UMAP to the reducedDims slot with the name "osca_UMAP" so we do not confused it with the author's UMAP coords.
sce <- runUMAP(sce, dimred = 'PCA', name = "osca_UMAP")

# When we want to access a specific item (here, "osca_UMAP") in the reducedDims element of the SingleCellExperiment object, we can use the function plotReducedDim with the argument dimred.

gg <- plotReducedDim(sce, dimred = "osca_UMAP", color_by = "cell_type", text_by = "cell_type") 
gg + theme(legend.title=element_blank()) + scale_color_manual(values=c(cell.color))
```

Now with a TSNE.
```{r, TSNE2, fig.height = 4, fig.width = 6.5}
# Run TSNE
set.seed(00101001101)

# runTSNE() stores the t-SNE coordinates in the reducedDims
sce <- runTSNE(sce, dimred = 'PCA', name = "osca_TSNE")

gg <- plotReducedDim(sce, dimred = "osca_TSNE", color_by = "cell_type", text_by = "cell_type") 
gg + theme(legend.title=element_blank()) + scale_color_manual(values=c(cell.color))
```


## Reference 

Ximerakis et al. (2023)
Heterochronic parabiosis reprograms the mouse brain transcriptome by shifting aging signatures in multiple cell types. \emph{Nat Aging} 3, 327â€“345. DOI:https://doi.org/10.1038/s43587-023-00373-6.



## View Session Info
```{r, sesh info}
sessionInfo()
```
