---
title: "Quick OSCA workflow on Parabiosis Mouse Data"
author: "Tram N. Nguyen"
date: "15 December, 2023"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Quick OSCA workflow on Parabiosis Mouse Data}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  tidy.opts = list(width.cutoff = 100), tidy = TRUE
)
```

```{r, setup, message=FALSE}
library(MouseAgingData)
library(devtools)
library(DropletUtils)
library(edgeR)
library(Seurat)
library(scran)
library(ggplot2)
library(SingleCellExperiment)
library(AnnotationHub)
```

# 1. Introduction

This vignette performs a simple analysis of the Ximerakis et al.
([2023](https://www.nature.com/articles/s43587-023-00373-6)) 10X Genomics
parabiosis dataset. The analysis follows the Basics of Single-Cell Analysis in
the [OSCA Bioconductor book Quick Start
Workflow](https://bioconductor.org/books/3.15/OSCA.intro/analysis-overview.html#quick-start-simple).
Steps in this vignette include: quality control, normalization, various forms of
dimensionality reduction, clustering into subpopulations, detection of marker
genes, and annotation of cell types.


# 2. Data Loading

### About the dataset

Parabiosis single cell RNA-seq (Ximerakis, Holton et al Nature Aging 2023)
105,329 cells, 31 cell types
8 OX, 8 YX, 7 YY, 9 YO, 7 OO, 11 OY
20905 features

Processed:
(ALL) reduced to: OLG, OPC, GLUT, GABA, EC, ASC
Aging OX v YX; RJV: OY v OX; AGA YO v YX

<br>

### Make the SingleCellExperiment Object. 

Load cell barcodes, genes, and matrix files using "DropletUtils".

```{r, Data Loading}

library(MouseAgingData)

# getWebData(format = "tsv", destination = "~/Desktop/", url = "https://ftp.ncbi.nlm.nih.gov/geo/series/GSE222nnn/GSE222510/suppl/GSE222510_processed_barcodes.tsv.gz", header=F)

# Eventually we can load in data through our package

filepath = "~/Documents/Rubin Mouse Aging/data/parabiosis/"

# load in components
barcodes <- read.table(paste0(filepath, "raw_files/10X/barcodes.tsv"), header = FALSE, stringsAsFactors = FALSE)
genes <- read.table(paste0(filepath,"raw_files/10X/genes.tsv"), header = FALSE, stringsAsFactors = FALSE)
metadata <- read.table(paste0(filepath, "raw_files/filtered.metadata.txt"), header = TRUE, sep = "") #105329 2

# Put all together in SCE
sce <- read10xCounts(paste0(filepath,"raw_files/10X/"), col.names = FALSE)
sce
```

Do some quick QC to make sure the data loaded in correctly, and is what we expected.

```{r, Data check}
colData(sce) # Sample metadata - just the cell barcodes

rowData(sce) # DataFrame with 20905 genes ID -- matches the features in their paper

dim(sce)

head(counts(sce)) # The count matrix. Transcript for each gene per cell (column) -- NOT NORMALIZED

# Grab the animal columns to investigate
df = cbind(metadata$animal, metadata$animal_name)
df = as.data.frame(unique(df))
table(df$V2)
```


#### Add the metadata to the sce object
We can add metadata to describe the columns of our primary data, e.g., the samples or cells of our experiment. This is stored in the colData slot, a DataFrame object where rows correspond to cells and columns correspond to metadata fields, e.g., batch of origin, treatment condition

```{r, add metadata}
dim(metadata) # Info for each cell (row n=10k). In the counts assay cells are columns.
SummarizedExperiment::colData(sce) <- DataFrame(metadata)
```

View our new `SingleCellExperiment` object. We see there are 12 ColData names now.

```{r, add metadata cont}
sce
head(colData(sce))

# check that rows and columns are the same number of cells
stopifnot(identical(length(metadata$barcode), ncol(counts(sce)))) # Number of barcodes should match columns of assay
```


# 3. Quality control

Explore and visualize mitochondrial content and read count. Note: the authors have already removed low-quality cells and animals so we will skip this section in this vignette. For more details on their workflow, one can refer to the original article Ximerakis et al. ([2023](https://www.nature.com/articles/s43587-023-00373-6)). The OSCA workbook [OSCA Bioconductor book Quick Start
Workflow](https://bioconductor.org/books/3.15/OSCA.intro/analysis-overview.html#quick-start-simple) provides several examples of quality control steps as well. 


<br>

# 4. Normalization

```{r, normalize}

library(scran)
set.seed(101000110)
clusters <- quickCluster(sce)
sce2 <- computeSumFactors(sce, clusters=clusters)
sce2 <- logNormCounts(sce2)


head(logcounts(sce2))


```

# 5. Feature selection.

Looks like these steps are all done already even with the raw data available - subsetted to final number of gene features (n=20905).

<br>

```{r, feat select}
library(scran)
dec <- modelGeneVar(sce2)
hvg <- getTopHVGs(dec, prop=0.1)
```

# 6. PCA.

```{r, PCA, warning=F}
set.seed(1234)
<<<<<<< HEAD
sce <- runPCA(sce, ncomponents=25, subset_row=hvg) # Add PCA coords to sce object
=======
sce3 <- runPCA(sce2, ncomponents=25, subset_row=hvg)

>>>>>>> parent of 9d1d2a3 (added sce rds)
```

# 7. Clustering.

```{r, clustering}
library(bluster)
colLabels(sce3) <- clusterCells(sce3, use.dimred='PCA',
                               BLUSPARAM=NNGraphParam(cluster.fun="louvain"))
```


# 8. Visualization.



<<<<<<< HEAD
# Generate color map matching cell type to colors in publication
color.map.list_All_Lin=c(
  #"OPC","OLG","OEG",
  "olivedrab4", "olivedrab3", "olivedrab1",
  #"NSC","ARP","ASC",
  "royalblue4", "steelblue4", "steelblue1",
  #"EPC","HypEPC,"TNC","CPC",
  "lightgoldenrod4", "lightgoldenrod3", "lightgoldenrod2","gold",
  #,"NRP","ImmN","GABA", DOPA", "GLUT","CHOL","NendC"
  "darkmagenta", "purple3","mediumorchid3", "violetred3", "palevioletred", "violet","lightpink",
  #,"EC","PC","VSMC","Hb-VC","VLMC","ABC",
  "sienna4", "sienna3", "sienna1", "peru","peachpuff4","peachpuff3",
  #"MG","MAC","MNC","DC","NEUT","T_cell","NK", "B_Cell"
  "red4","red3","tomato3", "red1","tomato1","salmon3", "indianred2", "coral")

names(color.map.list_All_Lin)<-c("OPC","OLG","OEG",
                                  "NSC","ARP","ASC",
                                  "EPC", "HypEPC", "TNC", "CPC",
                                  "NRP","ImmN","GABA", "DOPA", "GLUT","CHOL","NendC",
                                  "EC","PC","VSMC","Hb_VC","VLMC","ABC",
                                  "MG","MAC","MNC","DC","NEUT","T_cell","NK", "B_cell")


gg <- plotUMAP(sce, color_by = "cell_type", text_by = "cell_type") 
gg + theme(legend.title=element_blank()) + scale_color_manual(values=c(color.map.list_All_Lin))

# remove ori.gent, sample_order
# animal_name switch name to "parabiosis category" to describe the treatment pairing
# change period to underscores - consistent.

TSNE2 <- reducedDim(sce, "TSNE")
reducedDim(sce, "TSNE2") <- TSNE2 # Add another component to redcuedDims
# reorder the list
reducedDims(sce) <- reducedDims(sce)[c("TSNE", "TSNE2", "UMAP")] # Can do this with assays as well.

# remove rowData gene names
```

This plot is a recreation of Fig. 2C from Ximerakis et al. 2023.


```{r, plot provided TSNE, fig.height = 4, fig.width = 6.5}

gg <- plotTSNE(sce, color_by = "cell_type", text_by = "cell_type") 
gg + theme(legend.title=element_blank()) + scale_color_manual(values=c(color.map.list_All_Lin))
=======
>>>>>>> parent of 9d1d2a3 (added sce rds)
```




Start with a UMAP.
```{r, UMAP, fig.height = 4, fig.width = 7}
sce3 <- runUMAP(sce3, dimred = 'PCA')

<<<<<<< HEAD
```{r, UMAP, fig.height = 4, fig.width = 6.5}

OSCA_UMAP <- runUMAP(sce, dimred = 'PCA')
reducedDims(sce_test) <- c(reducedDims(sce), list(OSCA_UMAP=OSCA_UMAP))


# When we want to access a specific item (here, "UMAP2") in the reducedDims element of the SingleCellExperiment object, we can use the function plotReducedDim with the argument dimred.

gg <- plotReducedDim(sce, dimred = "OSCA_UMAP", color_by = "cell_type", text_by = "cell_type") 
gg + theme(legend.title=element_blank()) + scale_color_manual(values=c(color.map.list_All_Lin))
=======
plotReducedDim(sce3, dimred="UMAP", colour_by="cell_type", text_by = "cell_type")
>>>>>>> parent of 9d1d2a3 (added sce rds)
```


Now, TSNE plot.

```{r, TSNE, fig.height = 4, fig.width = 7}
#snn.gr <- buildSNNGraph(sce3, use.dimred="PCA", k=25)
#colLabels(sce3) <- factor(igraph::cluster_walktrap(snn.gr)$membership)

#plotTSNE(sce3, colour_by="label", text_by = "cell_type")

# Run TSNE
set.seed(00101001101)

# runTSNE() stores the t-SNE coordinates in the reducedDims
# for re-use across multiple plotReducedDim() calls.
sce3 <- runTSNE(sce3, dimred="PCA")
plotReducedDim(sce3, dimred="TSNE", colour_by="cell_type", text_by = "cell_type")
```


## Reference 

Ximerakis et al. (2023)
Heterochronic parabiosis reprograms the mouse brain transcriptome by shifting aging signatures in multiple cell types. \emph{Nat Aging} 3, 327â€“345. DOI:https://doi.org/10.1038/s43587-023-00373-6.

<br>

## View Session Info
```{r, sesh info}
sessionInfo()
```
