---
title: "Quick OSCA workflow on Parabiosis Mouse Data"
author: "Tram N. Nguyen"
date: "11 December, 2023"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Quick OSCA workflow on Parabiosis Mouse Data}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  tidy.opts = list(width.cutoff = 100), tidy = TRUE
)
```

```{r, setup, message=FALSE}
library(MouseAgingData)
library(devtools)
library(DropletUtils)
library(edgeR)
library(Seurat)
library(scran)
library(ggplot2)
library(SingleCellExperiment)
library(AnnotationHub)
```

## Introduction

This vignette performs a simple analysis of the Ximerakis et al. ([2023](https://www.nature.com/articles/s43587-023-00373-6)) 10X Genomics 
parabiosis dataset. The analysis follows the Basics of Single-Cell Analysis in 
the [OSCA Bioconductor book Quick Start Workflow](https://bioconductor.org/books/3.15/OSCA.intro/analysis-overview.html#quick-start-simple). 
Steps in this vignette include: quality control, normalization, 
various forms of dimensionality reduction, clustering into subpopulations, 
detection of marker genes, and annotation of cell types. 


## Data Loading

#### Information about the dataset
Initial dataset - 158,767 cells with data for 21,876 genes
Parabiosis single cell RNA-seq (Ximerakis, Holton et al Nature Aging 2023)
105,329 cells, 31 cell types
8 OX, 8 YX, 7 YY, 9 YO, 7 OO, 11 OY
20905 features
Processed:
(ALL) reduced to: OLG, OPC, GLUT, GABA, EC, ASC
Aging OX v YX; RJV: OY v OX; AGA YO v YX

<br>

#### Make the SingleCellExperiment Object. 
Load barcodes, genes, and matrix files using DropletUtils

```{r, Data Loading}

library(MouseAgingData)

# getWebData(format = "tsv", destination = "~/Desktop/", url = "https://ftp.ncbi.nlm.nih.gov/geo/series/GSE222nnn/GSE222510/suppl/GSE222510_processed_barcodes.tsv.gz", header=F)

# Eventually we can load in data through our package

filepath = "~/Documents/Rubin Mouse Aging/data/parabiosis/"

# load in components
barcodes <- read.table(paste0(filepath, "raw_files/10X/barcodes.tsv"), header = FALSE, stringsAsFactors = FALSE)
genes <- read.table(paste0(filepath,"raw_files/10X/genes.tsv"), header = FALSE, stringsAsFactors = FALSE)
metadata <- read.table(paste0(filepath, "raw_files/filtered.metadata.txt"), header = TRUE, sep = "") #105329 2

# Put all together in SCE
sce <- read10xCounts(paste0(filepath,"raw_files/10X/"), col.names = FALSE)
sce
```

Do some quick QC to make sure the data loaded in correctly, and is what we expected.

```{r, Data check}
colData(sce) # Sample metadata - just the cell barcodes

rowData(sce) # DataFrame with 20905 genes ID -- matches the features in their paper

dim(sce)

head(counts(sce)) # The count matrix. Transcript for each gene per cell (column) -- NOT NORMALIZED

# Grab the animal columns to investigate
df = cbind(metadata$animal, metadata$animal_name)
df = as.data.frame(unique(df))
table(df$V2)
```


#### Add the metadata to the sce object
We can add metadata to describe the columns of our primary data, e.g., the samples or cells of our experiment. This is stored in the colData slot, a DataFrame object where rows correspond to cells and columns correspond to metadata fields, e.g., batch of origin, treatment condition

```{r, add metadata}
dim(metadata) # Info for each cell (row n=10k). In the counts assay cells are columns.
SummarizedExperiment::colData(sce) <- DataFrame(metadata)
```

View our new `SingleCellExperiment` object. We see there are 12 ColData names now.

```{r, add metadata cont}
sce
head(colData(sce))

# check that rows and columns are the same number of cells
stopifnot(identical(length(metadata$barcode), ncol(counts(sce)))) # Number of barcodes should match columns of assay
```

## Quality control

Explore and visualize mitochondrial content and read count.
<span style="color:blue">*Note:* this is just an example from the scRNAseq package.</span>

<br>

```{r, QC check, eval=FALSE, include=FALSE}
library(AnnotationHub)
library(scRNAseq)
sce.mam <- scRNAseq::BachMammaryData(samples="G_1")

library(scater)
rownames(sce.mam) <- uniquifyFeatureNames(
    rowData(sce.mam)$Ensembl, rowData(sce.mam)$Symbol)

ens.mm.v97 <- AnnotationHub()[["AH73905"]]
rowData(sce.mam)$SEQNAME <- mapIds(ens.mm.v97, keys=rowData(sce.mam)$Ensembl,
    keytype="GENEID", column="SEQNAME")

unfiltered <- sce.mam
is.mito <- rowData(sce.mam)$SEQNAME == "MT"
stats <- perCellQCMetrics(sce.mam, subsets=list(Mito=which(is.mito)))
qc <- quickPerCellQC(stats, percent_subsets="subsets_Mito_percent")
sce.mam <- sce.mam[,!qc$discard]
colData(unfiltered) <- cbind(colData(unfiltered), stats)
unfiltered$discard <- qc$discard

gridExtra::grid.arrange(
    plotColData(unfiltered, y="sum", colour_by="discard") + 
        scale_y_log10() + ggtitle("Total count"),
    plotColData(unfiltered, y="detected", colour_by="discard") + 
        scale_y_log10() + ggtitle("Detected features"),
    plotColData(unfiltered, y="subsets_Mito_percent", 
        colour_by="discard") + ggtitle("Mito percent"),
    ncol=2
)
```

```{r, QC plot2, eval=F, include=FALSE}
plotColData(unfiltered, x="sum", y="subsets_Mito_percent", 
    colour_by="discard") + scale_x_log10()

#Figure 12.2: Percentage of mitochondrial reads in each cell in the Bach mammary gland dataset compared to its total count. Each point represents a cell and is colored according to whether that cell was discarded.

```

```{r mito plots, fig.height = 4, fig.width = 7}
library(scRNAseq)
sce.zeisel <- ZeiselBrainData()

library(scater)
sce.zeisel <- aggregateAcrossFeatures(sce.zeisel, 
    id=sub("_loc[0-9]+$", "", rownames(sce.zeisel)))

library(org.Mm.eg.db)
rowData(sce.zeisel)$Ensembl <- mapIds(org.Mm.eg.db, 
    keys=rownames(sce.zeisel), keytype="SYMBOL", column="ENSEMBL")

unfiltered <- sce.zeisel
# The original authors of the study have already removed low-quality cells prior to data publication. Nonetheless, we compute some quality control metrics to check whether the remaining cells are satisfactory.

stats <- perCellQCMetrics(sce.zeisel, subsets=list(
    Mt=rowData(sce.zeisel)$featureType=="mito"))
qc <- quickPerCellQC(stats, percent_subsets=c("altexps_ERCC_percent", 
    "subsets_Mt_percent"))
sce.zeisel <- sce.zeisel[,!qc$discard]
colData(unfiltered) <- cbind(colData(unfiltered), stats)
unfiltered$discard <- qc$discard

gridExtra::grid.arrange(
    plotColData(unfiltered, y="sum", colour_by="discard") +
        scale_y_log10() + ggtitle("Total count"),
    plotColData(unfiltered, y="detected", colour_by="discard") +
        scale_y_log10() + ggtitle("Detected features"),
    plotColData(unfiltered, y="altexps_ERCC_percent",
        colour_by="discard") + ggtitle("ERCC percent"),
    plotColData(unfiltered, y="subsets_Mt_percent",
        colour_by="discard") + ggtitle("Mito percent"),
    ncol=2
)

```


<br>

## Normalization

```{r, normalize}

library(scran)
set.seed(101000110)
clusters <- quickCluster(sce)
sce2 <- computeSumFactors(sce, clusters=clusters)
sce2 <- logNormCounts(sce2)


head(logcounts(sce2))


```

## Feature selection.

Looks like these steps are all done already even with the raw data available - subsetted to final number of gene features (n=20905).

<br>

```{r, feat select}
library(scran)
dec <- modelGeneVar(sce2)
hvg <- getTopHVGs(dec, prop=0.1)
```

## PCA.

```{r, PCA}
set.seed(1234)
sce3 <- runPCA(sce2, ncomponents=25, subset_row=hvg)

```

## Clustering.

```{r, clustering}
library(bluster)
colLabels(sce3) <- clusterCells(sce3, use.dimred='PCA',
                               BLUSPARAM=NNGraphParam(cluster.fun="louvain"))
```


## Visualization.

Start with a UMAP.
```{r, UMAP, fig.height = 4, fig.width = 7}
sce3 <- runUMAP(sce3, dimred = 'PCA')

plotReducedDim(sce3, dimred="UMAP", colour_by="cell_type", text_by = "cell_type")
```


Now, TSNE plot.

```{r, TSNE, fig.height = 4, fig.width = 7}
#snn.gr <- buildSNNGraph(sce3, use.dimred="PCA", k=25)
#colLabels(sce3) <- factor(igraph::cluster_walktrap(snn.gr)$membership)

#plotTSNE(sce3, colour_by="label", text_by = "cell_type")

# Run TSNE
set.seed(00101001101)

# runTSNE() stores the t-SNE coordinates in the reducedDims
# for re-use across multiple plotReducedDim() calls.
sce3 <- runTSNE(sce3, dimred="PCA")
plotReducedDim(sce3, dimred="TSNE", colour_by="cell_type", text_by = "cell_type")
```


## Reference 

Ximerakis et al. (2023)
Heterochronic parabiosis reprograms the mouse brain transcriptome by shifting aging signatures in multiple cell types. \emph{Nat Aging} 3, 327â€“345. DOI:https://doi.org/10.1038/s43587-023-00373-6.

<br>

## View Session Info
```{r, sesh info}
sessionInfo()
```
